internal/semaphore/semaphore.go

```diff
@@ -0,0 +1,70 @@
+// Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+// SPDX-License-Identifier: MIT
+
+package semaphore
+
+import (
+	"log"
+)
+
+type Semaphore interface {
+	//Acquire a slot in the semaphore
+	Acquire()
+
+	//Release a slot in the semaphore.
+	Release()
+
+	//GetLimit returns semaphore limit.
+	GetLimit() int
+
+	//GetCount returns current number of occupied slots in semaphore.
+	GetCount() int
+
+	//Done closes down the semaphore
+	Done()
+}
+
+type semaphore struct {
+	slots chan struct{}
+}
+
+func NewSemaphore(limit int) Semaphore {
+	if limit < 0 {
+		log.Panic("E! Semaphore size limit must not be negative")
+	}
+
+	sem := &semaphore{
+		slots: make(chan struct{}, limit),
+	}
+
+	for i := 0; i < cap(sem.slots); i++ {
+		sem.slots <- struct{}{}
+	}
```

Comment:
I was mostly fine with what it was before where you started with an empty channel and pushed to it as files got monitored. Why did you invert it so that you allocate all of the "slots" ahead of time?

Reply:
There is a slightly different regarding performance between these twos when scaling up to passed 1000 tailers. Here is an benchmark testing 
```
BenchmarkNewSemaphore_Acquire_ReleaseUnderLimitSimple-12          19538026                60.56 ns/op            0 B/op          0 allocs/op
BenchmarkNewSemaphore_Acquire_ReleaseUnderLimit-12                     8692            124471 ns/op               0 B/op          0 allocs/op
BenchmarkNewSemaphore_Acquire_ReleaseOverLimit-12                     32360             37397 ns/op               0 B/op          0 allocs/op
BenchmarkOldSemaphore_Acquire_ReleaseUnderLimitSimple-12        28253834                41.13 ns/op            0 B/op          0 allocs/op
BenchmarkOldSemaphore_Acquire_ReleaseUnderLimit-12                   9055            125437 ns/op               0 B/op          0 allocs/op
BenchmarkOldSemaphore_Acquire_ReleaseOverLimit-12                   31078             38852 ns/op               0 B/op          0 allocs/op
```